src/LowerParallelTasks.cpp

```diff
@@ -0,0 +1,519 @@
+
+    /** Codegen a call to do_parallel_tasks */
+    struct ParallelTask {
+        Stmt body;
+        struct SemAcquire {
+            Expr semaphore;
+            Expr count;
+        };
+        std::vector<SemAcquire> semaphores;
+        std::string loop_var;
+        Expr min, extent;
+        Expr serial;
+        std::string name;
+    };
+
+    using IRMutator::visit;
+
+    Stmt visit(const For *op) override {
+        const Acquire *acquire = op->body.as<Acquire>();
+
+        if (op->for_type == ForType::Parallel ||
+            (op->for_type == ForType::Serial &&
+             acquire &&
+             !expr_uses_var(acquire->count, op->name))) {
+            return do_as_parallel_task(op);
+        }
+        return IRMutator::visit(op);
+    }
+
+    Stmt visit(const Acquire *op) override {
+        return do_as_parallel_task(op);
+    }
+
+    Stmt visit(const Fork *op) override {
+        return do_as_parallel_task(op);
+    }
+
+    Stmt rewrite_parallel_tasks(const std::vector<ParallelTask> &tasks) {
+        Stmt body;
+
+        Closure closure;
+        for (const auto &t : tasks) {
+            Stmt s = t.body;
+            if (!t.loop_var.empty()) {
+                s = LetStmt::make(t.loop_var, 0, s);
+            }
+            s.accept(&closure);
+        }
+
+        // The same name can appear as a var and a buffer. Remove the var name in this case.
+        for (auto const &b : closure.buffers) {
+            closure.vars.erase(b.first);
+        }
+
+        int num_tasks = (int)(tasks.size());
+        std::vector<Expr> tasks_array_args(2);
+        tasks_array_args[0] = Call::make(type_of<halide_parallel_task_t *>(), Call::declare_struct_type, {StringImm::make("halide_parallel_task_t"), 0}, Call::PureIntrinsic);
+        tasks_array_args[1] = num_tasks;
+
+        std::string closure_name = unique_name("parallel_closure");
+        Expr closure_struct_allocation = AllocateClosure(closure_name, closure);
+        Expr closure_struct = Variable::make(Handle(), closure_name);
+
+        Expr result;
+        for (int i = 0; i < num_tasks; i++) {
+            ParallelTask t = tasks[i];
+
+            // Analyze the task body
+            class MayBlock : public IRVisitor {
+                using IRVisitor::visit;
+                void visit(const Acquire *op) override {
+                    result = true;
+                }
+
+            public:
+                bool result = false;
+            };
+
+            // TODO(zvookin|abadams): This makes multiple passes over the
+            // IR to cover each node. (One tree walk produces the min
+            // thread count for all nodes, but we redo each subtree when
+            // compiling a given node.) Ideally we'd move to a lowering pass
+            // that converts our parallelism constructs to Call nodes, or
+            // direct hardware operations in some cases.
+            // Also, this code has to exactly mirror the logic in get_parallel_tasks.
```

Comment:
At the least, we should probably add a similar comment to get_parallel_tasks() to reduce the chances the code diverges.